social_engagement/tasks.py

```diff
@@ -67,20 +69,26 @@ def task_handle_change_after_signal(user_id, course_id, param, increment=True, i
     """
     factor = items if increment else -items
     social_metric_points = get_social_metric_points()
-
-    score, _ = StudentSocialEngagementScore.objects.get_or_create(
-        user__id=user_id,
-        course_id=CourseKey.from_string(course_id),
-    )
-    if isinstance(param, dict):
-        for key, value in param.items():
-            score.score += social_metric_points.get(key, 0) * factor * value
-
-            previous = getattr(score, key, 0)
-            setattr(score, key, previous + value * factor)
+    course_key = CourseKey.from_string(course_id)
+    try:
+        user = User.objects.get(id=user_id)
```

Comment:
We can [prefetch the related](https://docs.djangoproject.com/en/1.11/ref/models/querysets/#prefetch-related) `studentsocialengagementscore_set` here so that getting the `StudentSocialEngagementScore` doesn't require another DB call:

```python
from django.db.models import Prefetch
...
try:
    user = User.objects.prefetch_related(
        Prefetch('studentsocialengagementscore_set', queryset=StudentSocialEngagementScore.objects.filter(course_id=course_key))
    ).get(id=user_id)
except User.DoesNotExist:
    ...
else:
    with transaction.atomic():
        # This is not another DB call; it was prefetched and is available in Python.
        score = user.studentsocialengagementscore_set.first() or StudentSocialEngagementScore(user=user, course_id=course_key)
        ...
```

There's likely a way of doing the calculations below as well really efficiently (literally all in a single database call) with [F expressions](https://docs.djangoproject.com/en/2.1/ref/models/expressions/#f-expressions) but I'm good with just going with this for now since this is only about fixing a bug.

Reply:
@UmanShahzad, good idea!
I'll do it as a part of `BB-492`.